snapshot/git/gitfiler/pool.go

```diff
@@ -0,0 +1,89 @@
+package gitfiler
+
+import (
+	"github.com/scootdev/scoot/snapshot/git/repo"
+)
+
+// repoAndError holds a repo and an error (just so we can pass it across a channel)
+type repoAndError struct {
+	repo *repo.Repository
+	err  error
+}
+
+// NewRepoPool creates a new RepoPool populated with existing repos and a getter that can get new ones
+func NewRepoPool(getter RepoGetter, repos []*repo.Repository, doneCh chan struct{}) *RepoPool {
+	freeList := make([]repoAndError, len(repos))
+	for i, v := range repos {
+		freeList[i] = repoAndError{repo: v}
+	}
+
+	p := &RepoPool{
+		getter:    getter,
+		releaseCh: make(chan repoAndError),
+		reserveCh: make(chan repoAndError),
+		doneCh:    doneCh,
+		freeList:  freeList,
+	}
+	go p.loop()
+	return p
+}
+
+// RepoPool lets clients Get a Repository for an operation, and then Release it when done
+// RepoPool can create a new Repository by using a supplied RepoGetter.
+// New repos can be added by the client by Release'ing a new Repository
+// Cf. sync's Pool
+type RepoPool struct {
```

Comment:
I'm not following, don't we only want to log requests/actions? If nothing is happening without us making a request then what is there to log?

In terms of background work, we use the loop construct to enable concurrent requests for internal state change, but I don't think that actually enables parallelism. So if we have no internal state to protect and we're already not parallel, just do the work in the foreground?

This is kind of moot though since full checkouts are super expensive. That makes me wonder if we should improve parallelism for  new clones.